### DISCUSSION TITLE
Whether to use async def or sync def when a request needs to access a shared object with lock

### DISCUSSION THREAD

**Discussion by zch-cc** [fastapi/fastapi#11418] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
No code just a conceptual question.
```

### Description

Hi I have a quick question on whether I should use async def or sync def as an endpoint for my application and here is what I want to achieve. 
I have a shared object(Let's say a hashmap) for all the request. For each request, it either access the map to retrieve value or add item to the value and doing some heavy cpu work and return a response. Since multiple request can access the hashmap at the same time, I add a thread lock to all the operations. My question is: Should I use async def or sync def for the endpoint? I find out the literature that the async function does not work really well with thread lock.  But in the docs for fastapi, it said 
```If your application (somehow) doesn't have to communicate with anything else and wait for it to respond, use async def.```
I am a little bit confused. Another thing is my application is cpu-bound not io-bound so make function async will not likely improve the performance so I think I should use def as an endpoint? 

### Operating System

Linux

### FastAPI Version

0.89.1

### Pydantic Version

1.10.4

### Python Version

3.8.10

**Comment by YuriiMotov** [comment, marked as answer]
Mentioned phrase from docs is really confusing. It was already discussed (https://github.com/tiangolo/fastapi/discussions/11203). 

You are right. Since your endpoints are CPU-bound and lock is a blocking mechanism, you should use `def` (not `async def`).


You can actually use `async def`, but in this case you will have to:

1. insert some `await`s to your code to 'split up' long code and allow event loop switching between async tasks.
For example
```
for item in data:
    await asyncio.sleep(0)
    # your code
```
2. Find way how to check the lock without locking the event loop:
For example
```
while lock.check_no_block():  # Asynchronously wait until the lock is unlocked
    await asyncio.sleep(0)
```
